### DISCUSSION TITLE
How to print the value in jax

### DISCUSSION THREAD

**Discussion by jimmylihui** [jax-ml/jax#18835] [Q&A] [answered]
<img width="638" alt="jax" src="https://github.com/google/jax/assets/47742416/45771b3f-5f42-4e3b-8d42-ba1e1f5eeda8">
The print doesn't print the information of error. Does anyone know how to print in Jax?

**Comment by jakevdp** [comment]
Hi - thanks for the question!

In the future, please copy and paste the code as text rather than as a screenshot – you can put it in a block between triple back-tick-marks (```` ``` ````) to format it. It's also helpful if you construct a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the issue you're having, so we have to make fewer guesses about what might be causing the problem.

To answer your question, the call to `jax.debug.print` looks correct to me, and should print the value contained in `error[0]` to `stdout`. One guess: if you're running your code in a Jupyter notebook, the `stdout` stream may be displayed in the terminal from which the notebook was launched rather than in the output cell in the notebook.

Can you say more about how you're running your code, what you're actually seeing, and what you expect to see?

**Reply by jimmylihui** [reply]
Great thanks. Now my function works,
```python
@eqx.filter_jit
def Linear(x,y,parameter):
    alpha=0.001
    x=jnp.atleast_2d(x)
    #x[batch, input_dimension], y[batch, output dimension], parameter[input, output]
    forward=einsum('bi,io->bo',x,parameter)
    forward=sigmoid(forward)
    error=forward-y

    D=error*sigmoid_deriv(forward)
    # delta=D.dot(parameter.T)
    # a=x.T.dot(D)
    parameter+= -alpha*x.T.dot(D)

    return parameter,error

@functools.partial(jax.jit, static_argnames=( "model"))
def loop(parameter,input,output,model):
    error_list=[]
    for i in range(input.shape[0]):
        x=input[i:i+1,:]
        y=output[i:i+1,:]
        parameter,error=model(x,y,parameter)
        error_list.append(error)
        jax.debug.print("{error}", error=error[0])
    return parameter,error_list
```
Following is the printed result,
```
[-0.0501774]equential method
[-1.919223]
[-0.61971784]
[-0.217705]
[1.8954756]
[-1.6327919]
[2.8948383]
[0.21038806]
[1.2499006]
[1.7467844]
[-0.9805677]
[-0.19084752]
[-0.33154643]
[0.96192527]
[2.2491715]
```
That is weird even if I change learning rate alpha, error wouldn't change. Do you have any idea?

**Reply by jakevdp** [reply]
Don't forget back-ticks for code formatting - I edited your post for you.

**Reply by jakevdp** [reply]
It seems like the values are being printed – what is your question?

**Reply by jimmylihui** [reply]
```python
@eqx.filter_jit
def Linear(x,y,parameter):
    alpha=0.1
    x=jnp.atleast_2d(x)
    #x[batch, input_dimension], y[batch, output dimension], parameter[input, output]
    forward=einsum('bi,io->bo',x,parameter)
    forward=sigmoid(forward)
    error=forward-y

    D=error*sigmoid_deriv(forward)
    # delta=D.dot(parameter.T)
    a=x.T.dot(D)
    jax.debug.print("parameter before:{parameter}", parameter=parameter[0])
    parameter+= -alpha*x.T.dot(D)
    jax.debug.print("parameter after:{parameter}", parameter=parameter[0])
    return parameter,error
```

The result is shown below

```python
parameter before:[-1.7753183]
parameter after:[-1.7547464]
parameter before:[-1.7546065]
parameter after:[-1.7546065]
parameter before:[-1.7633415]
parameter after:[-1.7633415]
parameter before:[-1.7656941]
parameter after:[-1.7656941]
parameter before:[-1.7660894]
parameter after:[-1.7660894]
parameter before:[-1.8295524]
parameter after:[-1.8295524]
parameter before:[-1.8497633]
parameter after:[-1.8497633]
parameter before:[-1.8251902]
parameter after:[-1.8251902]
parameter before:[-1.8305788]
parameter after:[-1.8305788]
parameter before:[-1.8280289]
parameter after:[-1.8280289]
parameter before:[-1.7860788]
parameter after:[-1.7860788]
parameter before:[-1.798911]
parameter after:[-1.798911]
parameter before:[-1.7968287]
parameter after:[-1.7968287]
parameter before:[-1.7993668]
parameter after:[-1.7993668]
parameter before:[-1.7969431]
parameter after:[-1.7969431]
parameter before:[-1.7692602]
parameter after:[-1.7692602]
```

It seems backward propagation is not working.

**Reply by jakevdp** [reply]
If you want help with model correctness, we’ll need to see all the code. I’d suggest putting together a minimal reproducible example that demonstrates the behavior you’re seeing (see the link I shared in my initial response).

**Reply by jimmylihui** [reply]
```python
import argparse
import time
import functools
import jax
import jax.numpy as jnp
import equinox as eqx

@eqx.filter_jit
def Linear(x,y,parameter):
    alpha=0.1
    x=jnp.atleast_2d(x)
    #x[batch, input_dimension], y[batch, output dimension], parameter[input, output]
    forward=einsum('bi,io->bo',x,parameter)
    forward=sigmoid(forward)
    error=forward-y

    D=error*sigmoid_deriv(forward)
    # delta=D.dot(parameter.T)
    a=x.T.dot(D)
    jax.debug.print("parameter before:{parameter}", parameter=parameter[0])
    parameter+= -alpha*x.T.dot(D)
    jax.debug.print("parameter after:{parameter}", parameter=parameter[0])
    return parameter,error

@functools.partial(jax.jit, static_argnames=( "model"))
def loop(parameter,input,output,model):
    error_list=[]
    for i in range(input.shape[0]):
        x=input[i:i+1,:]
        y=output[i:i+1,:]
        parameter,error=model(x,y,parameter)
        error_list.append(error)
        # jax.debug.print("{error}", error=error[0])
    return parameter,error_list


key = jax.random.PRNGKey(seed)
key, *subkey = jax.random.split(key, 4)
x = jax.random.normal(subkey[0], (10, 2), dtype=dtype)
y=jax.random.normal(subkey[1], (10, 1), dtype=dtype)
parameter=jax.random.normal(subkey[2], (2,1), dtype=dtype)
model=linear
parameter,error=loop(parameter,inputs,outputs,model)
```
please try this.

**Reply by jakevdp** [reply, marked as answer]
`seed`, `dtype`, and `linear` are undefined in your code.

**Reply by jimmylihui** [reply]
I have found the issue. Thanks for your help.
